File: ts/packages/benchmarks/src/translationBench/public_datasets/DroidCall/eval/officialDroidCallGrader.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { PythonNumber } from "../../pythonLiteral.js";

const SCORER_REVISION = "3f7ba458bee480a86c602edff6cc7ec9cfd555db";
const NUMBER_LEXEME = "__pythonNumber";

export type DroidCallContractName =
    | "paper-described"
    | "released"
    | "typeagent-adjusted";
export type DroidCallMatchType = "strict" | "semantic" | "ignore";

export type DroidCallArgumentSpec = {
    required?: boolean;
    default?: unknown;
    match_type?: DroidCallMatchType;
};
export type DroidCallTool = {
    name: string;
    arguments: Record<string, DroidCallArgumentSpec>;
};
export interface DroidCallOfficialRow {
    response: readonly { name: string; arguments: Record<string, unknown> }[];
    answers: readonly { name: string; arguments: Record<string, unknown> }[];
}
export type DroidCallSemanticScorer = (left: string, right: string) => number;

export interface DroidCallContractScore {
    softAccuracy: number;
    accuracy: number;
    counts: Record<
        | "rows"
        | "perfectRows"
        | "correctArguments"
        | "totalArguments"
        | "functionCalls",
        number
    >;
    contract: {
        name: DroidCallContractName;
        scorerRevision: string;
        semanticScorer: "token-overlap" | "caller-supplied";
        semanticThreshold: number;
        softAccuracyAggregation: "function-call-mean" | "sample-mean";
        overrides: DroidCallOverride[];
    };
}

type DroidCallOverride = {
    tool: "ACTION_OPEN_DOCUMENT";
    argument: "mime_types";
    comparison: "presence-only";
};
type ValueRecord = Record<string, unknown>;
function isRecord(value: unknown): value is ValueRecord {
    return typeof value === "object" && value !== null && !Array.isArray(value);
}
function hasOwn(value: ValueRecord, key: string): boolean {
    return Object.prototype.hasOwnProperty.call(value, key);
}
function isFieldNone(value: unknown): boolean {
    return (
        value === null ||
        value === undefined ||
        (typeof value === "string" && value.trim().toLowerCase() === "none")
    );
}
type NumericValue = number | bigint;
function numberValue(value: unknown): NumericValue | undefined {
    if (typeof value === "number") return value;
    const source =
        value instanceof PythonNumber
            ? value.lexeme
            : isRecord(value) && Object.keys(value).length === 1
              ? value[NUMBER_LEXEME]
              : undefined;
    if (typeof source !== "string") return undefined;
    return /^[+-]?\d+$/.test(source) ? BigInt(source) : Number(source);
}
function numbersEqual(left: NumericValue, right: NumericValue): boolean {
    if (typeof left === "bigint" && typeof right === "number") {
        return Number.isSafeInteger(right) && BigInt(right) === left;
    }
    if (typeof left === "number" && typeof right === "bigint") {
        return Number.isSafeInteger(left) && BigInt(left) === right;
    }
    return left === right;
}

function tokenOverlap(left: string, right: string): number {
    const tokens = (value: string) =>
        new Set(value.toLowerCase().match(/[a-z0-9]+/g) ?? []);
    const a = tokens(left);
    const b = tokens(right);
    const union = new Set([...a, ...b]).size;
    return union === 0
        ? left.trim() === right.trim()
            ? 1
            : 0
        : [...a].filter((token) => b.has(token)).length / union;
}

function compare(
    left: unknown,
    right: unknown,
    matchType: DroidCallMatchType,
    threshold: number,
    semanticScorer: DroidCallSemanticScorer,
): boolean {
    if (matchType === "ignore") return true;
    if (isFieldNone(left) && isFieldNone(right)) return true;
    const leftNumber = numberValue(left);
    const rightNumber = numberValue(right);
    if (leftNumber !== undefined || rightNumber !== undefined) {
        return (
            leftNumber !== undefined &&
            rightNumber !== undefined &&
            numbersEqual(leftNumber, rightNumber)
        );
    }
    const same = (a: unknown, b: unknown): boolean =>
        compare(a, b, matchType, threshold, semanticScorer);
    if (isRecord(left) && isRecord(right)) {
        const keys = Object.keys(left);
        return (
            keys.length === Object.keys(right).length &&
            keys.every(
                (key) => hasOwn(right, key) && same(left[key], right[key]),
            )
        );
    }
    if (Array.isArray(left) && Array.isArray(right)) {
        return (
            left.length === right.length &&
            left.every((item) =>
                right.some((candidate) => same(item, candidate)),
            ) &&
            right.every((item) =>
                left.some((candidate) => same(item, candidate)),
            )
        );
    }
    if (typeof left !== "string" || typeof right !== "string")
        return left === right;
    if (matchType === "strict") {
        return left.trim().toLowerCase() === right.trim().toLowerCase();
    }
    if (isFieldNone(left) || isFieldNone(right)) return false;
    return semanticScorer(left, right) > threshold;
}

function settings(contract: DroidCallContractName) {
    if (contract === "paper-described") {
        return {
            semanticThreshold: 0.75,
            aggregation: "function-call-mean" as const,
            mimePresenceOnly: false,
        };
    }
    if (contract === "released" || contract === "typeagent-adjusted") {
        return {
            semanticThreshold: 0.85,
            aggregation: "sample-mean" as const,
            mimePresenceOnly: contract === "typeagent-adjusted",
        };
    }
    throw new Error(`Unknown DroidCall scoring contract: ${contract}`);
}

type DroidCallSettings = ReturnType<typeof settings>;

function scoreArgument(
    apiName: string,
    name: string,
    spec: DroidCallArgumentSpec,
    answerArguments: ValueRecord,
    responseArguments: ValueRecord,
    config: DroidCallSettings,
    semanticScorer: DroidCallSemanticScorer,
): boolean {
    const answerHas = hasOwn(answerArguments, name);
    const responseHas = hasOwn(responseArguments, name);
    if (
        config.mimePresenceOnly &&
        apiName === "ACTION_OPEN_DOCUMENT" &&
        name === "mime_types"
    ) {
        return answerHas && responseHas;
    }
    if (!answerHas && !responseHas) return true;
    if (spec.required === true && !answerHas) return false;
    return compare(
        answerHas ? answerArguments[name] : (spec.default ?? null),
        responseHas ? responseArguments[name] : (spec.default ?? null),
        spec.match_type ?? "strict",
        config.semanticThreshold,
        semanticScorer,
    );
}

function scoreCall(
    answer: DroidCallOfficialRow["answers"][number],
    response: DroidCallOfficialRow["response"][number] | undefined,
    api: DroidCallTool,
    config: DroidCallSettings,
    semanticScorer: DroidCallSemanticScorer,
): { correct: number; total: number; missing: boolean } {
    const arguments_ = Object.entries(api.arguments);
    if (response === undefined) {
        return { correct: 0, total: arguments_.length, missing: true };
    }
    let correct = 0;
    for (const [name, spec] of arguments_) {
        if (
            scoreArgument(
                api.name,
                name,
                spec,
                answer.arguments,
                response.arguments,
                config,
                semanticScorer,
            )
        ) {
            correct++;
        }
    }
    return { correct, total: arguments_.length, missing: false };
}

function scoreRow(
    row: DroidCallOfficialRow,
    apiByName: ReadonlyMap<string, DroidCallTool>,
    config: DroidCallSettings,
    semanticScorer: DroidCallSemanticScorer,
): {
    score: number;
    correct: number;
    total: number;
    callSoftTotal: number;
    callCount: number;
} {
    const responseByName = new Map(
        row.response.map((response) => [response.name, response]),
    );
    let correct = 0;
    let total = 0;
    let failed = false;
    let callSoftTotal = 0;
    let callCount = 0;
    for (const answer of row.answers) {
        const api = apiByName.get(answer.name);
        if (api === undefined) {
            throw new Error(`Unknown API '${answer.name}'`);
        }
        const call = scoreCall(
            answer,
            responseByName.get(answer.name),
            api,
            config,
            semanticScorer,
        );
        failed ||= call.missing;
        correct += call.correct;
        total += call.total;
        callSoftTotal += call.missing
            ? 0
            : call.total === 0
              ? 1
              : call.correct / call.total;
        callCount++;
    }
    const score = failed ? 0 : total === 0 ? 1 : correct / total;
    return { score, correct, total, callSoftTotal, callCount };
}

export function scoreDroidCallContract(
    rows: readonly DroidCallOfficialRow[],
    apis: readonly DroidCallTool[],
    contract: DroidCallContractName,
    semanticScorer: DroidCallSemanticScorer = tokenOverlap,
): DroidCallContractScore {
    const config = settings(contract);
    const apiByName = new Map(apis.map((api) => [api.name, api]));
    let rowSoftTotal = 0;
    let callSoftTotal = 0;
    let callCount = 0;
    let perfectRows = 0;
    let correctArguments = 0;
    let totalArguments = 0;

    for (const row of rows) {
        const scored = scoreRow(row, apiByName, config, semanticScorer);
        rowSoftTotal += scored.score;
        callSoftTotal += scored.callSoftTotal;
        callCount += scored.callCount;
        if (Math.abs(scored.score - 1) < 1e-6) perfectRows++;
        correctArguments += scored.correct;
        totalArguments += scored.total;
    }
    const rowCount = rows.length;
    return {
        softAccuracy:
            config.aggregation === "function-call-mean" && callCount > 0
                ? callSoftTotal / callCount
                : rowCount > 0
                  ? rowSoftTotal / rowCount
                  : 0,
        accuracy: rowCount > 0 ? perfectRows / rowCount : 0,
        counts: {
            rows: rowCount,
            perfectRows,
            correctArguments,
            totalArguments,
            functionCalls: callCount,
        },
        contract: {
            name: contract,
            scorerRevision: SCORER_REVISION,
            semanticScorer:
                semanticScorer === tokenOverlap
                    ? "token-overlap"
                    : "caller-supplied",
            semanticThreshold: config.semanticThreshold,
            softAccuracyAggregation: config.aggregation,
            overrides: config.mimePresenceOnly
                ? [
                      {
                          tool: "ACTION_OPEN_DOCUMENT",
                          argument: "mime_types",
                          comparison: "presence-only",
                      },
                  ]
                : [],
        },
    };
}

export class DroidCallContractGrader {
    public constructor(
        _legacyScriptPath?: string,
        private readonly semanticScorer: DroidCallSemanticScorer = tokenOverlap,
    ) {}

    public async score(
        rows: readonly DroidCallOfficialRow[],
        apis: readonly DroidCallTool[],
        contract: DroidCallContractName,
    ): Promise<DroidCallContractScore> {
        return scoreDroidCallContract(
            rows,
            apis,
            contract,
            this.semanticScorer,
        );
    }

    public async close(): Promise<void> {}
}

File: ts/packages/benchmarks/src/translationBench/public_datasets/DroidCall/eval/officialDroidCallGrader.py
#!/usr/bin/env python3
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

"""DroidCall paper, released-code, and TypeAgent-adjusted scorers."""

import json
import re
import sys


_semantic_scorer = None
_semantic_scores = {}


def is_field_none(value):
    return value is None or (
        isinstance(value, str) and value.strip().lower() == "none"
    )


def decode_number_lexemes(value):
    if isinstance(value, list):
        return [decode_number_lexemes(item) for item in value]
    if isinstance(value, dict):
        if set(value) == {"__pythonNumber"}:
            source = value["__pythonNumber"]
            return int(source) if re.fullmatch(r"[+-]?\d+", source) else float(source)
        return {key: decode_number_lexemes(item) for key, item in value.items()}
    return value


def collect_semantic_pairs(left, right, match_type, pairs):
    if match_type != "semantic" or is_field_none(left) or is_field_none(right):
        return
    if type(left) is not type(right):
        return
    if isinstance(left, dict):
        if len(left) != len(right):
            return
        for key, value in left.items():
            if key in right:
                collect_semantic_pairs(value, right[key], match_type, pairs)
    elif isinstance(left, list):
        if len(left) != len(right):
            return
        for left_item in left:
            for right_item in right:
                collect_semantic_pairs(left_item, right_item, match_type, pairs)
    elif isinstance(left, str):
        pairs.add((left, right))


def prepare_semantic_scores(pairs):
    global _semantic_scorer
    missing = [pair for pair in pairs if pair not in _semantic_scores]
    if not missing:
        return
    if _semantic_scorer is None:
        from bert_score import BERTScorer

        _semantic_scorer = BERTScorer(lang="en")
    _, _, scores = _semantic_scorer.score(
        [pair[0] for pair in missing],
        [pair[1] for pair in missing],
    )
    for pair, score in zip(missing, scores):
        _semantic_scores[pair] = float(score)


def deep_compare(left, right, match_type="strict", semantic_threshold=0.85):
    if match_type == "ignore":
        return True
    if is_field_none(left) and is_field_none(right):
        return True
    if type(left) is not type(right):
        return False
    if isinstance(left, dict):
        if len(left) != len(right):
            return False
        return all(
            key in right
            and deep_compare(value, right[key], match_type, semantic_threshold)
            for key, value in left.items()
        )
    if isinstance(left, list):
        if len(left) != len(right):
            return False
        return all(
            any(deep_compare(a, b, match_type, semantic_threshold) for b in right)
            for a in left
        ) and all(
            any(deep_compare(a, b, match_type, semantic_threshold) for a in left)
            for b in right
        )
    if isinstance(left, str):
        if match_type == "strict":
            return left.strip().lower() == right.strip().lower()
        return _semantic_scores[(left, right)] > semantic_threshold
    if isinstance(left, int):
        return left == right
    return False


def resolved_arguments(answer, response, api):
    for name, spec in api["arguments"].items():
        answer_has = name in answer["arguments"]
        response_has = name in response["arguments"]
        if not answer_has and not response_has:
            continue
        if spec["required"] and not answer_has:
            continue
        default = spec.get("default")
        yield (
            answer["arguments"].get(name, default),
            response["arguments"].get(name, default),
            spec.get("match_type", "strict"),
        )


def score_payload(payload, contract_name):
    if contract_name == "paper-described":
        semantic_threshold = 0.75
        aggregation = "function-call-mean"
        mime_presence_only = False
    elif contract_name == "released":
        semantic_threshold = 0.85
        aggregation = "sample-mean"
        mime_presence_only = False
    elif contract_name == "typeagent-adjusted":
        semantic_threshold = 0.85
        aggregation = "sample-mean"
        mime_presence_only = True
    else:
        raise ValueError(f"Unknown DroidCall scoring contract: {contract_name}")

    apis = {item["name"]: item for item in payload["apis"]}
    for row in payload["rows"]:
        for response in row["response"]:
            response["arguments"] = decode_number_lexemes(response["arguments"])
    pairs = set()
    prepared = []
    for row in payload["rows"]:
        response_map = {
            item.get("name", ""): item
            for item in row["response"]
            if isinstance(item, dict) and isinstance(item.get("name", ""), str)
        }
        prepared.append(response_map)
        for answer in row["answers"]:
            response = response_map.get(answer["name"])
            if response is None:
                continue
            for left, right, match_type in resolved_arguments(
                answer, response, apis[answer["name"]]
            ):
                collect_semantic_pairs(left, right, match_type, pairs)
    prepare_semantic_scores(pairs)

    row_soft_total = 0.0
    call_soft_total = 0.0
    call_count = 0
    perfect_rows = 0
    correct_arguments = 0
    total_arguments = 0
    for row, response_map in zip(payload["rows"], prepared):
        row_correct = 0
        row_total = 0
        for answer in row["answers"]:
            api = apis[answer["name"]]
            response = response_map.get(answer["name"])
            call_correct = 0
            call_total = 0
            if response is None:
                call_total = len(api["arguments"])
                row_total += call_total
                call_soft_total += 1.0 if call_total == 0 else 0.0
                call_count += 1
                continue
            for name, spec in api["arguments"].items():
                answer_has = name in answer["arguments"]
                response_has = name in response["arguments"]
                if (
                    mime_presence_only
                    and api["name"] == "ACTION_OPEN_DOCUMENT"
                    and name == "mime_types"
                ):
                    if answer_has and response_has:
                        row_correct += 1
                        call_correct += 1
                    row_total += 1
                    call_total += 1
                    continue
                if not answer_has and not response_has:
                    row_correct += 1
                    call_correct += 1
                    row_total += 1
                    call_total += 1
                    continue
                if spec["required"] and not answer_has:
                    row_total += 1
                    call_total += 1
                    continue
                default = spec.get("default")
                if deep_compare(
                    answer["arguments"].get(name, default),
                    response["arguments"].get(name, default),
                    spec.get("match_type", "strict"),
                    semantic_threshold,
                ):
                    row_correct += 1
                    call_correct += 1
                row_total += 1
                call_total += 1
            call_soft_total += (
                1.0 if call_total == 0 else call_correct / call_total
            )
            call_count += 1
        row_score = 1.0 if row_total == 0 else row_correct / row_total
        row_soft_total += row_score
        if abs(row_score - 1.0) < 1e-6:
            perfect_rows += 1
        correct_arguments += row_correct
        total_arguments += row_total
    row_count = len(payload["rows"])
    soft_accuracy = (
        call_soft_total / call_count
        if aggregation == "function-call-mean"
        else row_soft_total / row_count
    )
    overrides = []
    if mime_presence_only:
        overrides.append(
            {
                "tool": "ACTION_OPEN_DOCUMENT",
                "argument": "mime_types",
                "comparison": "presence-only",
            }
        )
    return {
        "softAccuracy": soft_accuracy,
        "accuracy": perfect_rows / row_count,
        "counts": {
            "rows": row_count,
            "perfectRows": perfect_rows,
            "correctArguments": correct_arguments,
            "totalArguments": total_arguments,
            "functionCalls": call_count,
        },
        "contract": {
            "name": contract_name,
            "scorerRevision": "3f7ba458bee480a86c602edff6cc7ec9cfd555db",
            "bertScore": "0.3.13",
            "transformers": "4.48.1",
            "semanticThreshold": semantic_threshold,
            "softAccuracyAggregation": aggregation,
            "overrides": overrides,
        },
    }


def main():
    if "--jsonl" in sys.argv:
        for line in sys.stdin:
            try:
                payload = json.loads(line)
                contract_name = payload.pop("contract", "released")
                print(
                    json.dumps(score_payload(payload, contract_name)), flush=True
                )
            except Exception as error:
                print(json.dumps({"error": str(error)}), flush=True)
        return
    payload = json.load(sys.stdin)
    contract_name = payload.pop("contract", "released")
    print(json.dumps(score_payload(payload, contract_name)))


if __name__ == "__main__":
    main()
